### ISSUE TITLE
Allow plugins to hook into ratings events

### ISSUE THREAD

**Issue by lazka** [quodlibet/quodlibet#963]
Original [issue 963](https://code.google.com/p/quodlibet/issues/detail?id=963) created by nick.boultbee on 2012-05-07T09:55:08.000Z:

It would be good for plugins to be able to subscribe to ratings events i.e. 
1. when user changes the rating of a song (or potentially album one day - see Issue #544, or maybe artist somehow too), signals are sent with the rating chosen
2. When a song is loaded, or periodically, any subscribed plugins would be able to somehow influence or override the rating coming from the QL library. This would require some thought especially for the case of multiple providers, as well as caching and timeouts.

The prime candidates for this are scrobblers - last.fm provides a track.love [1] and track.unlove [2] API commands  and I imagine there are others like this.

[1] http://www.last.fm/api/show/track.love
[2] http://www.last.fm/api/show/track.unlove

**Comment by cauerego** [comment]
just please don't mix it up QL ratings with last.fm track-love, as discussed in #2381! ;P

**Comment by LoveIsGrief** [comment]
Hi,

I'd like to start working on this issue simply to make #927 cleaner. In order to do so, I would just like to send out ratings events whenever they are changed in a track.  
Without having looked at the code, I'm assuming there's a listener either at track level or GUI component level where I can trigger an event to be forwarded to plugins.

May I start working on that sub-issue in that manner?

**Comment by LoveIsGrief** [comment]
Looking at this, there is already a `plugin_on_changed` event handler that is called when a song is rated or otherwise modified. The only problem is that one doesn't know which attribute has been modified. Therefore I think the better solution to track which attribute changed would have to be introduced.

I'm sure that would be a bigger change, so I'm not gonna tackle this now unless there's a precise idea or feedback.

**Comment by bassstorm** [comment]
> Looking at this, there is already a `plugin_on_changed` event handler that is called when a song is rated or otherwise modified. The only problem is that one doesn't know which attribute has been modified. Therefore I think the better solution to track which attribute changed would have to be introduced.

Agree, maybe just supplying an attribute change set along with affected song(s) would be a good deal for start.
Spent a couple of hours today how events/plugins work, not very successfully though...
